### PR TITLE
Avoid querying with secondary columns which might fetch and lock additional rows

### LINKED ISSUE #16977: CRDB optimization: Optimize selects targeting the primary key or unique keys
### Description

With concurrent updates to the tables for UserSession and SingleUseObject, there are a lot of rolled back transactions due to failed serialization. 

When using optimistic locking on CRDB, the select statements around UserSession use the ID of the object, and add more conditions around the type of session (offline or not), and if the entity is expired. This might be one of the causes.

```
    select
        jpauserses0_.id as id1_40_,
        jpauserses0_.metadata as metadata7_40_,
        jpauserses0_.version as version11_40_
    from
        kc_user_session jpauserses0_
    where
        jpauserses0_.realmId=?
        and jpauserses0_.offline=?
        and jpauserses0_.id=?
        and (
            jpauserses0_.expiration>1675951377993
            or jpauserses0_.expiration is null
        ) for update
            of jpauserses0_
```

Queries around SingleUseObject don't use the primary key; this might be another cause. This could possibly be refactored to have the objectId as the primary key.See #13989 for a discussion why it is the way it is today (JPA storage can't store a non-UUID as a primary key with the current setup), and how this can be changed.

```
    select
        jpasingleu0_.id as id1_31_,
        jpasingleu0_.metadata as metadata4_31_,
        jpasingleu0_.version as version6_31_
    from
        kc_single_use_obj jpasingleu0_
    where
        jpasingleu0_.objectKey=?
        and (
            jpasingleu0_.expiration>1675951378891
            or jpasingleu0_.expiration is null
        )
```

The details of this issue contain now an experiment to see if by avoiding the suffixes on the select those serializations can be avoided. Issue #16976 should be implemented first so see if this approach unblock the issue. 

Efforts might be futile for features like "maximum number of user sessions per user in a realm", as those will run a broader query. 

Other queries for children like this might still interfere, but this remains to be seen:

```
    select
        clientsess0_.fk_root as fk_root6_21_0_,
        clientsess0_.id as id1_21_0_,
        clientsess0_.id as id1_21_1_,
        clientsess0_.metadata as metadata4_21_1_,
        clientsess0_.fk_root as fk_root6_21_1_,
        clientsess0_.version as version5_21_1_
    from
        kc_client_session clientsess0_
    where
        clientsess0_.fk_root=?
```

```
    select
        notes0_.fk_root as fk_root4_32_0_,
        notes0_.id as id1_32_0_,
        notes0_.id as id1_32_1_,
        notes0_.name as name2_32_1_,
        notes0_.fk_root as fk_root4_32_1_,
        notes0_.value as value3_32_1_
    from
        kc_single_use_obj_note notes0_
    where
        notes0_.fk_root=?
```

At the beginning of the task the list of SQL queries during login for both tables should be recorded to see if it to implement what's suggested here.

### Motivation

Performance is very bad on CRDB as a lot of transactions are retried for https://github.com/keycloak/keycloak-benchmark/issues/203. This is all about avoiding transactions to be rolled back due to problems with serializability.

### Details

Suggested changes to the JpaMapKeycloakTransaction: When the query contains an ID, use the ID to fetch the entity, and post-process the result with the other criteria.

In the case of UserSession, those queries would return for the happy path no value anyway.

A similar approach has been applied here to leverage that pessimistic locking is only available for query-by-ID on HotRot (kudos to @mhajas), see `#pessimisticQueryRead()`: 

https://github.com/keycloak/keycloak/pull/16619/files#diff-2cd58c3c33d4f93a5ff9b8262ffe66ebc3f375e1c7a3d519652a8e293d50bf5dR222-R265

### PR DESCRIPTION
**Pull request by ahus1**
This resolves problematic locking queries databases running on SERIALIZABLE isolation level like CockroachDB

Closes #16977

### CODE REVIEW

**Comment by ahus1** [comment]
I found that with the expiration query as part of the query, CRDB would sometimes use that index to retrieve the rows, which would then lock a lot more rows than expected. This might have been due to it being very restrictive when a lot of expired sessions are in the table. 

Once I changed this, I didn't see any transaction serialization issues on the single use objects any more.

**Comment by ahus1** [comment]
@sguilhen - this takes two approaches to retrieve the user session by its ID and the single use object via its objectKey: 

* For the user sessions, fetch by ID when possible and restrict the results further by adding a condition to the returned object

* Similar approach for the objectKey which used to be queried with the expiry. Assuming that we'd clean up the expired entries regularly, there shouldn't be many expired entries there anyway. So it should be ok to remove them after the fact. 

Once this is approved, I'll have a look at a very minimal cleanup process for expired entities.

**Comment by hmlnarik** [comment]
The idea looks great, thank you!

It seems that the implementation should have been done in the physical rather than logical layer. The change should not necessarily reflect say in a file store or even Postgres with the same impact. Could it be kept local to the JPA implementation, or even apply to CRDB only?

**Comment by ahus1** [comment]
@hmlnarik - I'd argue that "use the ID to query the store when you have the ID" is valid for all stores, so it would be ok to have it on the logical level when it comes to `MapUserSessionProvider`.

It will also benefit PostgreSQL (simpler queries). I don't know how HotRod manages this, I assume it would also benefit from a simple get instead of a full query. 

With JPA I saw that it ran one fewer query against the DB when deleting a user session - this is due to the fact that `delete(queryParameters)` first calls `em.remove()` for cached entities, and then runs the query against the DB to delete potentially others.

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapStorage.java [hunk 1]
```diff
@@ -181,7 +181,10 @@ public Stream<E> read(QueryParameters<M> queryParameters) {
         }
 
         JpaPredicateFunction<RE> predicateFunc = mcb.getPredicateFunc();
-        if (this.isExpirableEntity) {
+        if (this.isExpirableEntity && (queryParameters.getLimit() != null || queryParameters.getOffset() != null)) {
+            // only when using pagination exclude expired entities in the query directrly
```
Comment (sguilhen) [inline model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapStorage.java:185]:
```suggestion
            // only when using pagination exclude expired entities in the query directly
```

**Review by sguilhen — COMMENTED** [review]
Thanks, @ahus1, I also like the changes proposed here.

@hmlnarik I agree with @ahus1  that other stores also benefit from simpler searches. I would keep the changes in the physical layer unless there's a strong reason not to do it, which I can't see atm.

**Comment by hmlnarik** [comment]
> @hmlnarik - I'd argue that "use the ID to query the store when you have the ID" is valid for all stores, so it would be ok to have it on the logical level when it comes to MapUserSessionProvider.

We share the same view here, the ID should be used, and it should be used with highest precedence. The sad part is that datastore apparently picks slower evaluation regardless.

The aim of the logical layer is to provide the physical layer the full condition to enable efficient evaluation of the condition by that datastore. In the current code, the physical layer translates it to `ID = ? AND REALM_ID = ? AND OFFLINE_FLAG = ?` which causes the database to pick a slower path; the new code makes it `ID = ?` and the other two conditions are checked in Keycloak code.

The condition for this to be fast is that the objects that are retrieved by the query are small, and I also see the practical benefit. I only prefer this to be an exceptional case and for that reason ask for a comment why the provider does this extra filtering in java code instead of the datastore.

Then I'm +1 on including this change.

**Comment by ahus1** [comment]
Thank you @sguilhen and @hmlnarik - I pushed a change which adds the comment, and further simplifies the delete, see https://github.com/keycloak/keycloak/compare/56822c87642f284431d1bc73fac3810d76d722b5..435a8161b1ed55650cd69e6cb47b655cbf9a8a27

I'll ask for a new review once the tests complete all green.

**Review by hmlnarik — APPROVED** [review]
Thank you @ahus1 ! Approving the update and per prior approval by @sguilhen

**Pull request merged**